### PR TITLE
STARTGATE

### PR DESCRIPTION
**Pull request by RustyBamboo**
As requested, here it is.
Has worked on all the bags I've tried so far, however I can see many cases this will fail.

Overview:
- Finds mean and stddev of color. Then thresholds plus/minus deviations
- Contours, skeleton, and geometry checks
- Python mission is meh
- In Works: Better algorithm for perception...

### CODE REVIEW

File: command/sub8_missions/missions/start_gate.py
Comment (mattlangford) [inline command/sub8_missions/missions/start_gate.py]: You should set up the vision proxy, these 6 lines then become
```python
yield sub.vision_proxies.start_gate.start()
yield sub.vision_proxies.start_gate.get_pose()
```

File: command/sub8_missions/missions/start_gate.py
Comment (mattlangford) [inline command/sub8_missions/missions/start_gate.py]: Won't we be at the surface when this mission is started?

File: command/sub8_missions/missions/start_gate.py
Comment (mattlangford) [inline command/sub8_missions/missions/start_gate.py]: Does it go through?

File: perception/sub8_perception/src/sub8_perception/start_gate.cc
Comment (mattlangford) [inline perception/sub8_perception/src/sub8_perception/start_gate.cc]: You should use the rectified image

File: perception/sub8_perception/src/sub8_perception/start_gate.cc
Comment (mattlangford) [inline perception/sub8_perception/src/sub8_perception/start_gate.cc]: I think `gate_line_buffer.capactiy() - 1` would be better here, unless the 29 is a high precision machine learned number that optimally solves the problem.

File: perception/sub8_perception/src/sub8_perception/start_gate.cc
Comment (mattlangford) [inline perception/sub8_perception/src/sub8_perception/start_gate.cc]: Make arbitrary numbers params

**Review by mattlangford — APPROVED** [review]
Do it work?
I'll let David look at it as well

File: perception/sub8_perception/src/sub8_perception/start_gate.cc
Comment (RustyBamboo) [reply perception/sub8_perception/src/sub8_perception/start_gate.cc]: Yes, I spoke with David and have ideas for smarter gate detection.

File: command/sub8_missions/missions/start_gate.py
Comment (RustyBamboo) [reply command/sub8_missions/missions/start_gate.py]: Movement distances are arbitrary values... So I'm working on a way to get a 3D pose estimate and dynamically figure out distance

File: command/sub8_missions/missions/start_gate.py
Comment (RustyBamboo) [reply command/sub8_missions/missions/start_gate.py]: Alright, will look into that

**Comment by RustyBamboo** [comment]
The gate detection works fairly consistently on a lot of the bags, but I don't think it's robust enough... So I've been trying to find better ways to do this. I haven't had a chance to run the python mission, and will hopefully improve it with pose estimations.

**Comment by sentree** [comment]
Do we have any good TRANSDEC bags of the start gate? If so, have you had successful runs with these? Also, seeing as the date of our next pool day remains an uncollapsed probabilistic wave function, I would recommend seeing if Ken can task someone on his team with modeling the gate for the simulator. That should give you a bit more insight than just testing on bags.

**Comment by RustyBamboo** [comment]
I've been testing on a few bags from the TRANSDEC as well with fairly positive results. I'm trying to find bags that contain brighter colors (such as having buoys nearby) that might skew the distribution.

Good idea, I'll speak to Ken.

File: command/sub8_missions/missions/start_gate.py
Comment (DSsoto) [inline command/sub8_missions/missions/start_gate.py]: Consider making the code below part of a class and having the above as data members.

File: perception/sub8_perception/CMakeLists.txt
Comment (DSsoto) [inline perception/sub8_perception/CMakeLists.txt]: C++ files should have extensions .cpp or .hpp, if you see any other files that violate this feel free to rename them.

File: perception/sub8_perception/CMakeLists.txt
Comment (DSsoto) [inline perception/sub8_perception/CMakeLists.txt]: You might not want to use GLOB, especially for just a couple of files. If a source file is added or removed but the CMakeLists is not changed, then it will not update and the build will fail.

GLOB is usually for gathering files that match an pattern, not for an explicit file. Try setting a variable called START_GATE_SOURCES instead.

File: perception/sub8_perception/CMakeLists.txt
Comment (DSsoto) [inline perception/sub8_perception/CMakeLists.txt]: You probably don't need to link to PCL libs, and we're not using bfd or dl anymore.

File: perception/sub8_perception/nodes/start_gate_node.cc
Comment (DSsoto) [inline perception/sub8_perception/nodes/start_gate_node.cc]: Specify what node or this will not be useful.

File: perception/sub8_perception/src/sub8_perception/start_gate.cc
Comment (DSsoto) [inline perception/sub8_perception/src/sub8_perception/start_gate.cc]: This could have a more descriptive name than distance.

File: perception/sub8_perception/src/sub8_perception/start_gate.cc
Comment (DSsoto) [inline perception/sub8_perception/src/sub8_perception/start_gate.cc]: Use modern c++ man! 
```
for(auto& rect : minRect)
{
  rectanglePoints.push_back(std::vector<cv::Point2f>(rect.tl(), rect.br()));
}

```

File: perception/sub8_perception/src/sub8_perception/start_gate.cc
Comment (DSsoto) [inline perception/sub8_perception/src/sub8_perception/start_gate.cc]: You should really look into using ROSCameraStream, it makes this trivial and handles a lot of the boilerplate for you and provides you with an image buffer in case you need temporal information. It also contains all the camera info stuff you need.

You can use it [like this](https://github.com/uf-mil/mil_common/blob/great_merge/perception/mil_vision/nodes/camera_stream_demo.cpp): `mil::ROSCameraStream<cv::Vec3b> left_cam_stream(nh, history_length);` @

File: perception/sub8_perception/src/sub8_perception/start_gate.cc
Comment (DSsoto) [inline perception/sub8_perception/src/sub8_perception/start_gate.cc]: `done` has an undefined value, as does the variable `max` below. You should always initialize your variables when you declare them unless you're sure that you won't be generating some undefined behavior.

File: perception/sub8_perception/src/sub8_perception/start_gate.cc
Comment (DSsoto) [inline perception/sub8_perception/src/sub8_perception/start_gate.cc]: You should comment this so that someone could be able to get the gist of what your algorithm is trying to do when glancing at the code.

File: perception/sub8_perception/src/sub8_perception/start_gate.cc
Comment (DSsoto) [inline perception/sub8_perception/src/sub8_perception/start_gate.cc]: Since you're not doing arithmetic with the indexing variable, you should use range-based for loops.

**Review by DSsoto — CHANGES_REQUESTED** [review]
Requesting some changes. Main points:
+ Make your code conform to the [ROS style guide](http://wiki.ros.org/CppStyleGuide) (2 spaces, no cuddle braces, underscores after class data members, etc.)
+ You should try to use more descriptive variable naming and add descriptive comments. Try to make the code easier to read for others.
+ Try to use more features of modern c++, they will make your code safer and more concise.
+ Avoid using magic constants

I'd like to see how your code runs on bags sometime, and I can help you get the full pose of the gate if the you were able to get the silhouette quadrilateral successfully.

File: perception/sub8_perception/CMakeLists.txt
Comment (kev-the-dev) [reply perception/sub8_perception/CMakeLists.txt]: *pushes back glasses* umm actually the [official ros style guide](http://wiki.ros.org/CppStyleGuide#Files) states that c++ header files should have the extension **.h**

Comment (DSsoto) [reply perception/sub8_perception/CMakeLists.txt]: Haha, let's not follow that part of the guide because it is stupid.

Comment (sentree) [reply perception/sub8_perception/CMakeLists.txt]: *Removes glasses dramatically* I kind of have to agree with @ironmig here, the style guide is clear and there's no good reason to ignore this part of it.

Comment (kingkevlar) [reply perception/sub8_perception/CMakeLists.txt]: _Breaks glasses_

Comment (samfisher47) [reply perception/sub8_perception/CMakeLists.txt]: *Gouges eyes out*

Comment (DSsoto) [reply perception/sub8_perception/CMakeLists.txt]: _Throws to the ground and kicks_ Me and @ironmig at the very least agreed that the good reason to ignore this part of it is because .h is for c files while .hpp is for c++.

Comment (mattlangford) [reply perception/sub8_perception/CMakeLists.txt]: Let's actually move to no header files, they're for the weak.

Comment (DSsoto) [reply perception/sub8_perception/CMakeLists.txt]: Re-write this in assembly, compiled languages are for the weak.

Comment (sentree) [reply perception/sub8_perception/CMakeLists.txt]: If we all wrote in assembly, there would be no compilation errors and we would get better performance. I'm for this.

File: command/sub8_missions/missions/start_gate.py
Comment (DSsoto) [inline command/sub8_missions/missions/start_gate.py]: This seems arbitrary since you don't know at what depth you're starting, what's the reason?

File: command/sub8_missions/missions/start_gate.py
Comment (DSsoto) [inline command/sub8_missions/missions/start_gate.py]: `FOUND_START_GATE` isn't actually a constant.

Can you make speed a ros param?

File: command/sub8_missions/missions/start_gate.py
Comment (DSsoto) [inline command/sub8_missions/missions/start_gate.py]: > A "New Class" is the recommended way to create a class in modern Python.

`class StartGateMission(object):`

File: command/sub8_missions/missions/start_gate.py
Comment (DSsoto) [inline command/sub8_missions/missions/start_gate.py]: Dead code

**Review by DSsoto — COMMENTED** [review]
I'll be ready to merge after these minor comments are addressed.

Could you add a unit test to this for your next PR?

File: perception/sub8_perception/src/sub8_perception/start_gate.cpp
Comment (DSsoto) [inline perception/sub8_perception/src/sub8_perception/start_gate.cpp]: This needs to be image_**rect**_color

File: perception/sub8_perception/src/sub8_perception/start_gate.cpp
Comment (DSsoto) [inline perception/sub8_perception/src/sub8_perception/start_gate.cpp]: `ROS_INFO_STREAM("Deleted size: " << s)` You shouldn't just use cout from code running on the robot.

File: perception/sub8_perception/src/sub8_perception/start_gate.cpp
Comment (DSsoto) [inline perception/sub8_perception/src/sub8_perception/start_gate.cpp]: Instead of using #ifdef visualize and opening gui windows, it would be cleaner to publish visualization images.

File: perception/sub8_perception/src/sub8_perception/start_gate.cpp
Comment (DSsoto) [inline perception/sub8_perception/src/sub8_perception/start_gate.cpp]: This should be `ROS_INFO_STREAM("Deleted size: " << s)`

Don't use cout from ros nodes, so that we can control the verbosity of the output.

File: perception/sub8_perception/src/sub8_perception/start_gate.cpp
Comment (DSsoto) [inline perception/sub8_perception/src/sub8_perception/start_gate.cpp]: Instead of using defines for visualization, the cleaner way would be to publish debug images.

**Comment by DSsoto** [comment]
@RustyBamboo any updates on this? It would be cool to merge before pool day tomorrow.

**Comment by RustyBamboo** [comment]
I would like to see if the sonar makes a usable pointcloud for the gate and implement that into this mission.
I'll work on a unittest.

**Review by DSsoto — APPROVED** [review]

**Pull request merged**
